cuda/matrix/hybrid_kernels.cu

```diff
@@ -0,0 +1,436 @@
+
+
+namespace kernel {
+
+
+// Copied from coo_kernel.cu
+template <int subwarp_size = cuda_config::warp_size, typename ValueType,
+          typename IndexType>
+__device__ __forceinline__ void segment_scan(
+    const group::thread_block_tile<subwarp_size> &group, IndexType ind,
+    ValueType *__restrict__ val, bool *__restrict__ head)
+{
+#pragma unroll
+    for (int i = 1; i < subwarp_size; i <<= 1) {
+        const IndexType add_ind = group.shfl_up(ind, i);
+        ValueType add_val = zero<ValueType>();
+        if (threadIdx.x >= i && add_ind == ind) {
+            add_val = *val;
+            if (i == 1) {
+                *head = false;
+            }
+        }
+        add_val = group.shfl_down(add_val, i);
+        if (threadIdx.x < subwarp_size - i) {
+            *val += add_val;
+        }
+    }
+}
+
+
+/**
+ * The global function for counting nonzeros of COO.
+ * It is almost like COO spmv routine.
+ *
+ * @param nnz  the number of nonzeros in the matrix
+ * @param num_line  the maximum round of each warp
+ * @param val  the value array of the matrix
+ * @param row  the row index array of the matrix
+ * @param c  the output nonzeros per row
+ */
+template <int subwarp_size = cuda_config::warp_size, typename ValueType,
+          typename IndexType, typename SizeType>
+__global__ __launch_bounds__(default_block_size) void count_coo_row_nnz(
+    const size_type nnz, const size_type num_lines,
+    const ValueType *__restrict__ val, const IndexType *__restrict__ row,
+    SizeType *__restrict__ c)
+{
+    SizeType temp_val = 0;
+    const auto start = static_cast<size_type>(blockDim.x) * blockIdx.x *
+                           blockDim.y * num_lines +
+                       threadIdx.y * blockDim.x * num_lines;
+    size_type num = (nnz > start) * ceildiv(nnz - start, subwarp_size);
+    num = min(num, num_lines);
+    const IndexType ind_start = start + threadIdx.x;
+    const IndexType ind_end = ind_start + (num - 1) * subwarp_size;
+    IndexType ind = ind_start;
+    bool is_first_in_segment = true;
+    IndexType curr_row = (ind < nnz) ? row[ind] : 0;
+    const auto tile_block =
+        group::tiled_partition<subwarp_size>(group::this_thread_block());
+    for (; ind < ind_end; ind += subwarp_size) {
+        temp_val += ind < nnz && val[ind] != zero<ValueType>();
+        auto next_row =
+            (ind + subwarp_size < nnz) ? row[ind + subwarp_size] : row[nnz - 1];
+        // segmented scan
+        if (tile_block.any(curr_row != next_row)) {
+            is_first_in_segment = true;
+            segment_scan<subwarp_size>(tile_block, curr_row, &temp_val,
+                                       &is_first_in_segment);
+            if (is_first_in_segment) {
+                atomic_add(&(c[curr_row]), temp_val);
+            }
+            temp_val = 0;
+        }
+        curr_row = next_row;
+    }
+    if (num > 0) {
+        ind = ind_start + (num - 1) * subwarp_size;
```

Comment:
I thought one of the reasons for declaring `ind` outside the `for`-loop was so you could continue using its value after the `for`-loop stopped (with the value of `ind` that was too big for the `for`-loop).
If you want to explicitly write it here, please use `ind_end`, so it is clear that you resume the `for`-loop (which is also less error prone):
```suggestion
        ind = ind_end;
```